### ISSUE TITLE
`builtins.readFile` doesn't work with certain files 

### ISSUE THREAD

**Issue by btlvr** [NixOS/nix#3292]
**Describe the bug**
I am trying to use `builtins.readFile`. It breaks when I try to use it with certain files.

For example, I encounter the following error:
```
nix-repl> builtins.readFile "/sys/devices/virtual/dmi/id/product_family"
error: unexpected end-of-file
```

**To Reproduce**
1. Try to use `builtins.readFile` to read the contents of any file in `/sys/devices/virtual/dmi/id/`

**Expected behavior**
`builtins.readFile` should do what it's supposed to do

**Additional context**
This problem seems to be specific to Nix, since I have no problems reading the file with other tools, such as `cat`

**Comment by zimbatm** [comment]
The issue is that this file's reported size is larger than the content. Probably a quirk of that virtual filesystem.
```
$ stat /sys/devices/virtual/dmi/id/product_family
  File: /sys/devices/virtual/dmi/id/product_family
  Size: 4096      	Blocks: 0          IO Block: 4096   regular file
Device: 19h/25d	Inode: 1764        Links: 1
Access: (0444/-r--r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2019-12-27 12:48:36.324598248 +0100
Modify: 2019-12-27 12:48:36.324598248 +0100
Change: 2019-12-27 12:48:36.324598248 +0100
 Birth: -
$ wc -c /sys/devices/virtual/dmi/id/product_family
23 /sys/devices/virtual/dmi/id/product_family
```
prim_readFile ends-up calling the following function which tries to read the file until `fstat().st_size`

https://github.com/NixOS/nix/blob/b0cadf547bf2f796b550d7555b97761ff1d0a379/src/libutil/util.cc#L308-L318

**Comment by jappeace** [comment]
It's a problem as well for files encrypted with git-crypt.

**Comment by jappeace** [comment]
builtins.hashFile works, so I compared against the hash instead.

**Comment by tim-hilt** [comment]
> It's a problem as well for files encrypted with git-crypt. I tried setting up a simple git-crypt detection by adding a file to the encrypted folder with the simple string "no" in it.

Yeah, that sounds like it's also related to the file-length, mentioned by @zimbatm. I wanted to set up a modular config based on the different motherboards of my laptops (identified by `/sys/devices/virtual/dmi/id/product_name`) and it also (unsurprisingly) didn't work.

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
